### PR TITLE
Kernel Density Estimate

### PR DESCRIPTION
**Pull request by arkwave**
matlab transferEntropyKDE and mdKDE functions are done, tests written and passing on local machine.

### CODE REVIEW

**Comment by AceHao** [comment]
looks good to me,

**Comment by Christopher-Tennant** [comment]
This looks good. Two quick questions, (1) Math is a "generic" Python library, correct? (2) The large number of text files is required because you are testing many of the cases presented with the original toolbox?

**Comment by arkwave** [comment]
Sorry for the late reply. Yes, Math is a library native to python. The large number of text files is essentially because i ran the entire simulation given in the demoscript, and verified correctness to the same degree for all of them.

**Comment by stefanv** [comment]
Note that kernel density estimates are already implemented in SciPy, so you probably want to re-use those.

Also, please use the numpy documentation format to document functions.

File: infoflow/entropy_kde.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import numpy as np
+from infoflow import multidimkde as mdkde
+from math import log
+
+
+def kerneldensityestimation(source, target, timelagx, timelagy, n, bw_coeff):
```
Comment (stefanv) [inline infoflow/entropy_kde.py:6]: According to PEP8, function names should be in the form `kernel_density_estimation`.  That also goes for `timelag_x` and `timelag_y`.

File: infoflow/entropy_kde.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    :param source: source time series, 1-D np.ndarray
```
Comment (stefanv) [inline infoflow/entropy_kde.py:8]:
```
Parameters
----------
source : 1-D ndarray
    Source time series
etc.
```

File: infoflow/entropy_kde.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    :param target: target time series, 1-D np.ndarray
+    :param timelagx: time lag in x from present. Integer.
+    :param timelagy: time lag in y from present. Integer.
+    :param n: number of equally spaced points along each dimension where probabilities are to be estimated. Integer.
+    :param bw_coeff: multiplier that adjusts Gaussian bandwidth. Integer.
+    :return: transfer entropy in bits. float.
+    """
+
+    """ All code written here is based on the transfer entropy
```
Comment (stefanv) [inline infoflow/entropy_kde.py:17]: Place this into the docstring above.

"Use protected under GNU" is incorrect.  You can say something like "licensed under the GNU GPL v3 license".

File: infoflow/entropy_kde.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    sourcearr = source.flatten()
+    targetarr = target.flatten()
+
+    # populating source_pat, target_pat and target_t
+    source_pat = []
+    target_pat = []
+    target_t = []
+
+    for i in range(max(l+timelagx, k+timelagy)-1, min(len(sourcearr), len(targetarr))):
```
Comment (stefanv) [inline infoflow/entropy_kde.py:31]: You almost never want to loop over arrays in this fashion.

Read more about numpy array operations here: http://www.scipy-lectures.org/intro/numpy/index.html

File: infoflow/multidimkde.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+def multidimensionalkde(x, xi, bw_coeff):
+    """
+    :param x: np.ndarray, rows are observations and columns are dimensions.
+    :param xi: np.ndarray, coordinates at which joint probabilities are to be estimated.
+    :param bw_coeff: multiplier that adjusts the rule of thumb bandwidth. Integer.
+    :return: joint probability at xi
+    """
+
+    # assertions to ensure input is in right format.
+    assert isinstance(x, np.ndarray)
```
Comment (stefanv) [inline infoflow/multidimkde.py:14]: These tests should be in the form

```
if not isinstance(x, np.ndarray):
    raise ValueError("x must be ndarray")
```

File: infoflow/multidimkde.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    h = np.zeros(shape=(1, d))
+    for i in range(d):
+        h[0, i] = float(bw_coeff * 1.06 * n**(-1/5) *np.std(x[:, i],ddof=1))
+    # estimating probability at xi
+    p = np.zeros(shape=(np.shape(xi)[0], 1))
+    for k in range(len(p)):
+        for i in range(n):
+            prod = 1
+            for j in range(d):
+                u = float((xi[j] - x[i, j])/h[0][j])
```
Comment (stefanv) [inline infoflow/multidimkde.py:28]: You never want to do this type of for-loop on a numpy array.

File: infoflow/tests/test_entropy_kde.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# tests first case of matlab demo script.
+def test_basic():
+    x = np.genfromtxt(path + '/input_kde/entropykdeinputx.txt', delimiter="\t")
+    y = np.genfromtxt(path + '/input_kde/entropykdeinputy.txt', delimiter='\t')
+    n = 10
+    timelagx = 2
+    timelagy = 2
+    bw_coeff = 1
+    ans = kde.kerneldensityestimation(x, y, timelagx, timelagy, n, bw_coeff)
+    result = 0.798553577280728
```
Comment (stefanv) [inline infoflow/tests/test_entropy_kde.py:18]: These kinds of tests are very bad at hitting corner cases.  What you want is to figure out how your algorithm responds to all extreme inputs and parameters, and work out minimal test cases for those examples.  Working out examples by hand are advisable!

File: infoflow/tests/test_multidimensionalkde.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+from infoflow import multidimkde as multkde
+import numpy as np
+import os
+
+path = os.path.dirname(os.path.realpath(__file__))
+
+def test_basic():
+    """
+    Note about test: ans[0][0] - result ~ 10^-34. If system is chaotic, could cause errors. Double check
```
Comment (stefanv) [inline infoflow/tests/test_multidimensionalkde.py:9]: Floating point precision cannot give you better than 10^-16 anyway.

File: infoflow/multidimkde.py (hunk 6 above)
Comment (arkwave) [inline infoflow/multidimkde.py:28]: How would you recommend vectorizing this?

**Comment by arkwave** [comment]
Alright in light of the issues thus far (lack of testing coverage and the existence of KDE in scipy), I'll take some time after finals to re-write this function after understanding the math behind it. @stefanv should I close this pull request and make a new one later?

**Comment by stefanv** [comment]
You could do that, or you can simply keep pushing to this branch.  In the
end, you can flatten / rewrite your commits.  I'm happy to chat about
vectorization sometime if you'd like.
